include/rnp/rnp2.h

```diff
@@ -0,0 +1,234 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/*
+* Currently we use some definitions/enums from other others, including
+* - rnp_result_t
+*
+* It might be good to consolidate these definitions into one spot, or
+* else replicate them here in new enums and make this header freestanding.
+*/
+#include <rnp/rnp_def.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+/*
+* Opaque structures for key types
+*/
+typedef struct rnp_keyring_st *rnp_keyring_t;
+
+/**
+* Callback used for getting a passphrase.
+* @param app_ctx provided by application in rnp_keyring_open
+* @param pgp_context a descriptive string for what is being decrypted
+* @param pass to which the callback should write the returned
+* passphrase, NULL terminated.
+* @param pass_len the size of pass buffer
+* @return 0 on success, or any other value to stop decryption.
+*/
+typedef int (*rnp_passphrase_cb)(void *      app_ctx,
+                                 const char *pgp_context,
+                                 char        buf[],
+                                 size_t      buf_len);
+
+/* Operations on key rings */
+
+rnp_result_t rnp_keyring_open(rnp_keyring_t *   keyring,
+                              const char *      keyring_format,
+                              const char *      pub_path,
+                              const char *      sec_path,
+                              rnp_passphrase_cb cb,
+                              void *            app_ctx);
+
+rnp_result_t rnp_keyring_close(rnp_keyring_t keyring);
+
+/**
+* Load a public key into the keyring
+* @param key a pointer, on success will be assigned to the new object
+* @param key_bits the binary serialized key
+* @param key_len length of key_bits in bytes
+*/
+rnp_result_t rnp_insert_public_key(rnp_keyring_t keyring,
+                                   const char *  key_format,
+                                   const uint8_t key_bits[],
+                                   size_t        key_len);
+
+/**
+* Load an armored public key into the keyring
+* @param key a pointer, on success will be assigned to the new object
+* @param key the PGP armored public key
+*/
+rnp_result_t rnp_insert_armored_public_key(rnp_keyring_t keyring, const char *key);
+
+/**
+* Generate a key and add it to the keyring
+*/
+rnp_result_t rnp_generate_private_key(rnp_keyring_t keyring,
```

Comment:
I'm kind of wondering if we would be better off just doing a quick implementation of https://github.com/riboseinc/rnp/issues/318 for this.

I started brainstorming on it a bit earlier today.
It would be pretty easy I think to pass some JSON to a key generation function. Would just need to merge in some defaults, check for unused object attributes (error), etc.

```javascript
// generate both a primary and subkey
{
"primary": {
    "type": "ECDSA",
    "curve": "secp256r1",
    "usage": ["sign"]
 },
 "subkey": {
    "type": "RSA",
    "length": 4096
 }
}

// generate just a subkey
{
"primary": {
    "grip": "...",
 },
 "subkey": {
    "type": "RSA",
    "length": 4096
 }
}
```

Anyways, it's worth considering I think. Because even with all these parameters here, we are still missing certain things (like key usage).

Also, since we have separated key generation from key protection, the passphrase parameters shouldn't be here.

Reply:
(Also I certainly wouldn't mind tackling this particular piece if you're thinking "I have enough work to do")

Reply:
Yeah I wasn't so happy with that. There are a lot of optional params and etc. So maybe JSON is the way to go here. If you want to take this part on feel free :)

Reply:
Ok I'll get started on that piece today, I think it'll be pretty quick.